sdk/go/auto/cmd.go

```diff
@@ -17,17 +17,237 @@ package auto
+		if err != nil {
+			return nil, err
+		}
+		newOpts.Root = path.Join(home, ".pulumi", "versions", newOpts.Version.String())
+	}
+
+	return newOpts, nil
+}
+
+type pulumiCommand struct {
+	version semver.Version
+	command string
+}
+
+// NewPulumiCommand creates a new PulumiCommand instance that uses a local CLI
+// installation. PulumiCommandOptions can be used to configure things like the
+// the installation root or to validate that the version meets a minimum
+// requirement. By default the pulumi binary found in $PATH is used.
+func NewPulumiCommand(opts *PulumiCommandOptions) (PulumiCommand, error) {
+	if opts == nil {
+		opts = &PulumiCommandOptions{}
+	}
+	command := "pulumi"
+	if opts.Root != "" {
+		command = path.Join(opts.Root, "bin", "pulumi")
+	}
+
+	cmd := exec.Command(command, "version")
+	out, err := cmd.Output()
+	if err != nil {
+		return pulumiCommand{}, fmt.Errorf("failed to run `pulumi version`: %w", err)
+	}
+	currentVersion := strings.TrimSpace(string(out))
+	min := minimumVersion
+	if opts.Version.GT(min) {
+		min = opts.Version
+	}
+	version, err := parseAndValidatePulumiVersion(min, currentVersion, opts.skipVersionCheck)
+	if err != nil {
+		return pulumiCommand{}, err
+	}
+
+	return pulumiCommand{
+		version: version,
+		command: command,
+	}, nil
+}
+
+// InstallPulumiCommand downloads and installs the Pulumi CLI. By default the
+// CLI version matching the current SDK release is installed in
+// $HOME/.pulumi/versions/$VERSION. Use the option `PulumiCommandRoot` to
+// specify a different directory, and `PulumiCommandVersion` to install a
+// custom version.
+func InstallPulumiCommand(ctx context.Context, opts *PulumiCommandOptions) (
+	PulumiCommand,
+	error,
+) {
+	if opts == nil {
+		opts = &PulumiCommandOptions{}
+	}
+	opts, err := opts.withDefaults()
+	if err != nil {
+		return pulumiCommand{}, err
+	}
+	pulumi, err := NewPulumiCommand(opts)
+	if err == nil {
+		// Found an installation and it satisfies the version requirement
+		return pulumi, nil
+	}
+	if runtime.GOOS == "windows" {
+		if err := installWindows(ctx, opts.Version, opts.Root); err != nil {
+			return pulumiCommand{}, err
+		}
+	} else {
+		if err := installPosix(ctx, opts.Version, opts.Root); err != nil {
+			return pulumiCommand{}, err
+		}
+	}
+	return NewPulumiCommand(opts)
+}
+
+func downloadToTmpFile(ctx context.Context, url, filePattern string) (string, error) {
+	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
+	if err != nil {
+		return "", err
+	}
+	resp, err := httputil.DoWithRetry(req, http.DefaultClient)
+	if err != nil {
+		return "", fmt.Errorf("failed to download %s: %w", url, err)
+	}
+	defer resp.Body.Close()
+	scriptData, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return "", fmt.Errorf("failed to read response from %s: %w", url, err)
+	}
+	tmp, err := os.CreateTemp("", filePattern)
+	if err != nil {
+		return "", err
+	}
+	scriptPath := tmp.Name()
+	onErrorCleanup := func() {
+		os.Remove(scriptPath)
+	}
```

Comment:
So the more go way to do this is to use a named return, then you could just have a defer function to see if `returnErr != nil`.

Reply:
Yeah, I don't love named returns, but I suppose it's not to bad here https://github.com/pulumi/pulumi/pull/15049/commits/9d4afe18b2fbc2ebd8cda0fdf2ce79485ad8b119